File: ai/10/10.cpp
#include <bits/stdc++.h>
using namespace std;

/*
    a* algorithm
    f(n) = g(n) + h(n)
    f(n) -> total cost
    g(n) -> distance from start
    h(n) -> heuristic function
*/

signed main(){
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    int n, e;
    cin >> n >> e;
    vector<vector<array<int, 2>>> adj(n + 1);
    for (int i = 0; i < e; i++){
        int u, v, w;
        cin >> u >> v >> w;
        adj[u].push_back({v, w});
        adj[v].push_back({u, w});
    }
    vector<int> h(n + 1);
    for (int i = 1; i <= n; i++){
        cin >> h[i];
    }
    int start, end;
    cin >> start >> end;
    priority_queue<array<long long, 2>, vector<array<long long, 2>>, greater<array<long long, 2>>> pq;
    const long long M = 1e18;
    vector<long long> f(n + 1, M);
    vector<int> par(n + 1, -1);
    pq.push({h[start], start});
    f[start] = h[start];
    while (pq.size() > 0){
        auto fro = pq.top();
        pq.pop();
        int u = fro[1];
        if (f[u] < fro[0]) {
            continue;
        }
        for (auto to : adj[u]){
            int v = to[0], w = to[1];
            if (f[v] > w + h[v] + (f[u] - h[u])){
                par[v] = u;
                f[v] = w + h[v] + (f[u] - h[u]);
                pq.push({f[v], v});
            }
        }
    }
    vector<int> path;
    for (int i = end; i != -1; i = par[i]) {
        path.push_back(i);
    }
    reverse(path.begin(), path.end());
    cout << "Path: ";
    for (int i = 0; i < (int)path.size(); i++) {
        cout << path[i];
        if (i != path.size() - 1) {
            cout << " -> ";
        }
    }
    cout << "\nCost: " << f[end] << '\n';
    return 0;
}

File: ai/10/10.py
import heapq

# A* algorithm
# f(n) = g(n) + h(n)
# f(n) -> total cost
# g(n) -> distance from start

def a_star(n, e, edges, heuristics, start, end):
    adj = [[] for _ in range(n + 1)]
    for u, v, w in edges:
        adj[u].append((v, w))
        adj[v].append((u, w)) 

    M = float('inf')
    f = [M] * (n + 1)
    par = [-1] * (n + 1)
    pq = []
    heapq.heappush(pq, (heuristics[start], start))
    f[start] = heuristics[start]

    print("Traversal: ", end = " ")
    while pq:
        cost_u, u = heapq.heappop(pq)
        print(u, end = " ")
        if f[u] < cost_u:
            continue
        for v, w in adj[u]:
            new_cost = w + heuristics[v] + (f[u] - heuristics[u])
            if f[v] > new_cost:
                par[v] = u
                f[v] = new_cost
                heapq.heappush(pq, (f[v], v))

    path = []
    curr = end
    while curr != -1:
        path.append(curr)
        curr = par[curr]
    path.reverse()

    print("\nPath:", " -> ".join(map(str, path)))
    print("Cost:", f[end])

if __name__ == "__main__":
    n, e = map(int, input().split())
    edges = [tuple(map(int, input().split())) for _ in range(e)]
    heuristics = [0] + list(map(int, input().split()))
    start, end = map(int, input().split())
    a_star(n, e, edges, heuristics, start, end)

'''
5 7
1 2 1
1 3 4
2 3 2
2 5 5
2 4 12
3 5 2
4 5 3
7 6 2 1 0
1 5
Traversal:  1 3 5 2 3 5 4 4 
Path: 1 -> 2 -> 3 -> 5
Cost: 5
'''